backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/ui/web/order/BOMExploderCommand.java

```diff
@@ -130,17 +147,60 @@ public BOMExploderCommand(@Nullable final BOMUse bomToUse,
 			final OrderLineCandidate lineCandidate = initialCandidate.toBuilder()
 					.productId(bomLineProductId)
 					.attributes(attributes)
-					.qty(bomLineQty)
+					.qty(bomLineQtys)
 					.compensationGroupId(compensationGroupId)
-					.uomId(bomUomId)
 					.explodedFromBOMLineId(bomLineId)
 					.build();
 			result.addAll(this.toBuilder()
 					.initialCandidate(lineCandidate)
 					.build()
-					.execute());  // recurse
+					.execute0());  // recurse
 		}
 
 		return result;
 	}
+
+	private List<OrderLineCandidate> reduceCandidates(final List<OrderLineCandidate> result)
+	{
+		return result.stream()
+				.collect(ImmutableMap.toImmutableMap(
+						OrderLineCandidate::getProductId,
+						candidate -> candidate,
+						this::combine))
+				.values()
+				.asList();
+	}
+
+	private OrderLineCandidate combine(final OrderLineCandidate olCand1, final OrderLineCandidate olCand2)
+	{
+		final Quantity olCand1Qty = olCand1.getQty();
+		final Quantity olCand2Qty = olCand2.getQty();
+
+		if (olCand1Qty.signum() == 0)
+		{
+			return olCand2;
+		}
+		if (olCand1Qty.signum() < 0)
+		{
+			throw new AdempiereException("Invalid qty: " + olCand1Qty);
+		}
+
+		if (olCand2Qty.signum() == 0)
+		{
+			return olCand1;
+		}
+		if (olCand2Qty.signum() < 0)
+		{
+			throw new AdempiereException("Invalid qty: " + olCand2Qty);
+		}
+
+		if (!olCand1.getAttributes().equals(olCand2.getAttributes()))
+		{
+			throw new AdempiereException(msgBL.getTranslatableMsgText(MSG_ATTRIBUTES_NOT_MATCHING, olCand1.getProductId(), olCand2.getProductId()));
```

Comment:
u can pass MSG_ATTRIBUTES_NOT_MATCHING directly to AdempiereException, no need to ask msgBL.

see org.adempiere.exceptions.AdempiereException#AdempiereException(de.metas.i18n.AdMessageKey, java.lang.Object...)

Reply:
replaced.